node/node.go

```diff
@@ -391,46 +388,37 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, chainDBFactory shardc
 	return &node
 }
 
-// InitShardState initialize genesis shard state and update committee pub keys for consensus and drand
-func (node *Node) InitShardState(isGenesis bool) (err error) {
-	logger := utils.GetLogInstance().New("isGenesis", isGenesis)
+// InitShardState initialize shard state from latest epoch and update committee pub keys for consensus and drand
+func (node *Node) InitShardState() (err error) {
+	logger := utils.GetLogInstance().New()
 	getLogger := func() log.Logger { return utils.WithCallerSkip(logger, 1) }
 	if node.Consensus == nil {
+		// will exit the program
 		getLogger().Crit("consensus is nil; cannot figure out shard ID")
 	}
 	shardID := node.Consensus.ShardID
-	logger = logger.New("shardID", shardID)
-	getLogger().Info("initializing shard state")
 
 	// Get genesis epoch shard state from chain
-	genesisEpoch := big.NewInt(core.GenesisEpoch)
-	shardState, err := node.Beaconchain().GetShardState(genesisEpoch, nil)
-	if err != nil {
-		return ctxerror.New("cannot read genesis shard state").WithCause(err)
+	blockNum := node.Blockchain().CurrentBlock().NumberU64()
+	node.Consensus.SetMode(consensus.Listening)
+	epoch := core.ShardingSchedule.CalcEpochNumber(blockNum)
+	getLogger().Info("[InitShardState] Try To Get PublicKeys from database", "blockNum", blockNum, "shardID", shardID)
+	pubKeys := core.GetPublicKeys(epoch, shardID)
+	if len(pubKeys) == 0 {
+		return ctxerror.New(
+			"[InitShardState] PublicKeys is Empty, Cannot update public keys",
+			"shardID", shardID,
+			"blockNum", blockNum)
 	}
-	getLogger().Info("Successfully loaded epoch shard state")
 
-	// Update validator public keys
-	committee := shardState.FindCommitteeByID(shardID)
-	if committee == nil {
-		return ctxerror.New("our shard is not found in genesis shard state",
-			"shardID", shardID)
-	}
-	pubKeys := []*bls.PublicKey{}
-	for _, node := range committee.NodeList {
-		pubKey := &bls.PublicKey{}
-		pubKeyBytes := node.BlsPublicKey[:]
-		err = pubKey.Deserialize(pubKeyBytes)
-		if err != nil {
-			return ctxerror.New("cannot deserialize BLS public key",
-				"shardID", shardID,
-				"pubKeyBytes", hex.EncodeToString(pubKeyBytes),
-			).WithCause(err)
+	for _, key := range pubKeys {
+		if key.IsEqual(node.Consensus.PubKey) {
```

Comment:
don't we have a function to check isLeader?

Reply:
This is to check if your key is in the record, not check whether it's leader or not